File: .github/custom-queries/matlab/convert_to_sarif.m
% Load the results from the .mat file
load('code-analysis-results.mat', 'results');

% Initialize SARIF structure
sarif.version = "2.1.0";
sarif.runs = struct([]);

% Initialize run object
run.tool.driver.name = "Octave Static Code Analysis for MATLAB";
run.tool.driver.version = "1.0";
run.results = [];

% Populate SARIF results
for i = 1:length(results)
    filePath = results{i}{1};
    issues = results{i}{2};
    
    % Loop over each issue found in the file
    for j = 1:length(issues)
        % Create SARIF result entry for each issue
        sarifResult = struct();
        sarifResult.ruleId = "MATLABIssue"; % Identifier for the issue type
        sarifResult.level = "warning";
        sarifResult.message = struct('text', issues{j});
        
        % Specify the location
        sarifResult.locations = struct([]);
        sarifResult.locations(1).physicalLocation.artifactLocation.uri = filePath;
        sarifResult.locations(1).physicalLocation.region.startLine = 1; % Set startLine if available
        
        % Add result to the SARIF run
        run.results = [run.results, sarifResult];
    end
end

% Add the populated run to SARIF
sarif.runs = [sarif.runs, run];

% Save SARIF output to a JSON file
jsonText = jsonencode(sarif);
fid = fopen('code-analysis-results.sarif', 'w');
if fid == -1
    error('Could not create SARIF file');
end
fwrite(fid, jsonText, 'char');
fclose(fid);

File: .github/custom-queries/matlab/convert_to_sarif.py
import scipy.io
import json

# Load the .mat file
data = scipy.io.loadmat('code-analysis-results.mat')
results = data.get('results', [])

# SARIF template
sarif = {
    "version": "2.1.0",
    "runs": [
        {
            "tool": {
                "driver": {
                    "name": "MATLAB Code Analysis",
                    "informationUri": "https://github.com/username/repo",
                    "rules": [
                        {
                            "id": "MATLAB001",
                            "name": "MATLAB Code Issue",
                            "shortDescription": {"text": "Identified MATLAB Code Issue"},
                            "fullDescription": {"text": "This issue was identified in the MATLAB code."},
                            "defaultConfiguration": {"level": "warning"}
                        }
                    ]
                }
            },
            "results": []
        }
    ]
}

# Convert each entry in the results to SARIF format
for item in results:
    file_path = item[0][0]
    issues = item[1][0]
    for issue in issues:
        sarif_issue = {
            "ruleId": "MATLAB001",
            "message": {"text": issue[0]},
            "locations": [
                {
                    "physicalLocation": {
                        "artifactLocation": {"uri": file_path},
                    }
                }
            ]
        }
        sarif["runs"][0]["results"].append(sarif_issue)

# Write to SARIF file
with open('code-analysis-results.sarif', 'w') as f:
    json.dump(sarif, f, indent=2)
